File: R/tutorial/basic-examples/branch_flow.R
library(metaflow)

a <- function(self){
    self$var = 1
}

b <- function(self){
    self$var = 2
}

join <- function(self, inputs){
    print(paste("var in step a is", inputs[[1]]$var))
    print(paste("var in step b is", inputs[[2]]$var))
}

metaflow("BranchFlow") %>%
    step(step = "start",
         next_step = c("a", "b")) %>%
    step(step = "a", 
         r_function=a, 
         next_step="join") %>%
    step(step="b", 
         r_function=b, 
         next_step="join") %>%
    step(step="join", 
         r_function=join, 
         next_step="end",
         join=TRUE) %>%
    step(step="end") %>%
    run()

File: test/ux/core/flows/dag/branch_flow.py
import os

from metaflow import FlowSpec, step, project


@project(name="branch_flow")
class BranchFlow(FlowSpec):
    @step
    def start(self):
        self.execution_env = os.environ.get("KUBERNETES_SERVICE_HOST", "")
        self.next(self.branch_a, self.branch_b)

    @step
    def branch_a(self):
        self.value = "a"
        self.next(self.join)

    @step
    def branch_b(self):
        self.value = "b"
        self.next(self.join)

    @step
    def join(self, inputs):
        self.values = sorted([i.value for i in inputs])
        self.next(self.end)

    @step
    def end(self):
        pass


if __name__ == "__main__":
    BranchFlow()
